### DISCUSSION TITLE
how to create an immutable class with list field? recommendation needed.

### DISCUSSION THREAD

**Discussion by wenhoujx** [pydantic/pydantic#3956] [Question]
hello, i am from java world, and the unhashable builtin list python thing is a big surprise. I try to have a pydantic `BaseModel` subclass that has a `list` field, and claim it's `frozen`, but that still doesn't make the class hashable, b/c the list field is not hashable. I wonder what's the best approach here, i see a few: 
1. hand write my own `__hash__` method 
2. use some other types to replace `list`, even a 3-rd party one like from `pyrsistent`. 

i don't like any one of them, the first one requires me to write my own `hash` methods for a lot of classes, the second one demands a little learning curve for future developers. 

please advise what to do and how you approach this problem. thanks 

here is a runnable pytest code to illustrate the problem: 
```
from typing import List

from pydantic import BaseModel

class TestClass(BaseModel):
    friends: List[str]
    class Config:
        frozen = True


def test():
    test_obj = TestClass(friends=['a', 'b'])
    hash(test_obj)
```

output 
```
test1.py::test FAILED                                                    [100%]
test1.py:9 (test)
def test():
        test_obj = TestClass(friends=['a', 'b'])
>       hash(test_obj)

test1.py:12: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: unhashable type: 'list'
```

**Comment by boriel** [comment]
An _immutable list_ is, by definition, a `tuple`. Just use `Tuple[str, ...]` instead.
